feat: multiprotocol compatibility with client 11.00 (10x)

marcosvf132 opened the pull request:
![image](https://user-images.githubusercontent.com/66353315/190416668-7748a77c-d0c9-476b-8067-e9b4d01ce1fc.png)
# PLEASE READ THE PR DESCRIPTION BEFORE WRITING A COMMENT.

# First of all
This PR is still on draft since so many things must be tested and there are still changes that will be coming, like editing the /modules/ scripts and the store structure., so you probably will find debbugs related to Store, Blessings or Daily reward. Check for future commits on this branch.

# To-Do list
- [x] Debug when open the outfit window on god char.
- [x] Test all systems and 12x features.

# "I don't wan't to allow players to login with client 10x on my server"
If you wan't to block 10x players on your server, just leave the config.lua tag _allowOldProtocol_ set to default (false)
![image](https://user-images.githubusercontent.com/66353315/190415840-b0c96265-fc03-41f6-9add-17fa15e92c3a.png)

# How the login works
The client 10x doesn't need at all any AAC to login, it connects to the server directly, so no changes on AAC need to be done, but to keep login on 12x you must keep using a AAC. On client 10x you can only login using **USERNAME**, and on client 12x you can only login using **EMAIL**.

# How to test it
Just set the tag _allowOldProtocol_  on config.lua to true and login using a 11.00 client (Don't forget to update it to your server sprites/items data)
![image](https://user-images.githubusercontent.com/66353315/190415612-fefef48d-b470-4456-a1ed-cb2a4af1c433.png)

# Current behavior
You can only login with client 12.91. (The current protocol)
![desativado](https://user-images.githubusercontent.com/66353315/190414980-8fa51090-e805-48bf-88c6-b46205ee5374.png)

# With this PR (And the tag _allowOldProtocol_ set to true on config.lua)
![ativado](https://user-images.githubusercontent.com/66353315/190415144-297192fc-810e-499b-8f99-f72ef06c5a0a.png)
![login](https://user-images.githubusercontent.com/66353315/190415185-1345af73-e8d0-4366-8bb5-210619bf204b.png)

marcosvf132 committed:
Initial commit (1 of 2)

- Second part of the commit will be the /data/ changes.

mattyx14 commented:
It is a good contribution for those who like the old client... but of course omitting all the great improvements with respect to the modern client...

beats-dh commented:
Maybe it works for something: https://github.com/beats-dh/canary/commits/Old-Protocol

marcosvf132 committed:
Initial commit (2/2)

Data modules change.
Npc trade fix
Quest log fix.

marcosvf132 commented:
> It is a good contribution for those who like the old client... but of course omitting all the great improvements with respect to the modern client...

 Indeed it's a big downgrade when we are talking about 12x features. The majority of features are disabled or are working on the background of the server (The server-side features that you active on 12x client and login on 10x).

 On the near future i will make modules to make possible to use some of the new 12 functionalities, but of course we are limited to modal window and dialog boxes of the client.

But the option on config.lua gives the opportunity to maintain all the implemented features already running as they should.

flowbomba commented:
Hello in past versions of multiprotocol, there is a bug that was never fixed.
The player's online status changes when in exercise training in the 12+ client.

and that causes client 10 to crash, when an online friend enters exersice training.

I have not tested your commit, but I only mention that possible bug, in case you missed it

flowbomba commented:
Hello, I have been testing the multiprotocol, I hope my comment will be helpful to polish it.

if there is a debug when a player on your friends list does training weapon, since it changes the online status to blue in client 12.91.

and another debug in client 10, it is with the hireling icons, (npc from houses)

luiscaldera commented:
hello I made all the changes in the src and config.lua to add version 11 and without problems when compiling but I get this error when I try to enter with version 10 or 12 if it lets me connect without problems

print:
![image](https://user-images.githubusercontent.com/35881739/193430315-5f393894-deb8-4d22-b9c2-218042702373.png)

jaysoncynthia commented:
![image](https://user-images.githubusercontent.com/107809642/196390464-c1056180-3222-4f9a-aaa4-b5f877760c55.png)

**hello friends, I downloaded the commit to test, but when I try to enter the 12.91 client, it gives this error there, on the client 10 I can log in normally - when I put true it in config.lua**

beats-dh commented:
> ![image](https://user-images.githubusercontent.com/107809642/196390464-c1056180-3222-4f9a-aaa4-b5f877760c55.png)
> 
> **hello friends, I downloaded the commit to test, but when I try to enter the 12.91 client, it gives this error there, on the client 10 I can log in normally - when I put true it in config.lua**

Did you change the 12x client rsa key?
Can you log in with the 12x client with false in config.lua?

flowbomba commented:
hello, the system can no longer be tested due to changes in the source code, with the tier system, it interferes and gives an error. can you update? have problems in protocolgame.cpp

flowbomba commented:
please update :(

Nyedson commented:
Usando o client 10, com um player normal, ao lançar uma magia que tem o grupo "focus" (Exevo gran mas frigo, tera, flam, vis, utito tempo, utamo tempo) no cooldown o cliente dá debug e aparece isso no client:

![Sem título](https://user-images.githubusercontent.com/41634901/219208692-be6548e4-2f64-4884-a88d-a286d5b91930.png)

Só aqui?

Edit¹.: Retirei o focus do cooldown e não deu o debug, por isso sei que tem relação com ele.

dudantas commented:
> Usando o client 10, com um player normal, ao lançar uma magia que tem o grupo "focus" (Exevo gran mas frigo, tera, flam, vis, utito tempo, utamo tempo) no cooldown o cliente dá debug e aparece isso no client:
> 
> ![Sem título](https://user-images.githubusercontent.com/41634901/219208692-be6548e4-2f64-4884-a88d-a286d5b91930.png)
> 
> Só aqui?
> 
> Edit¹.: Retirei o focus do cooldown e não deu o debug, por isso sei que tem relação com ele.

A branch está desatualizada... Logo, não vai funcionar corretamente. Em breve devemos voltar a desenvolver este pull, vai testando e reportando os problemas que achar aqui nos comentários.

dudantas committed:
Merge branch 'main' into pr/516

dudantas committed:
add missing enum

github-actions[bot] committed:
Code format - (Clang-format)

flowbomba commented:
[ProtocolGame::onRecvFirstMessage] - RSA Decrypt

dudantas commented:
> [ProtocolGame::onRecvFirstMessage] - RSA Decrypt

You need to enable oldProtocol in config.lua

flowbomba commented:
Hello I have been testing, there are some old errors and I found a new one:

- The most critical error so far is that the outfit cannot be changed with client 10 give debug, therefore in otcv8 it does not open the outfits panel (it only works with client 12, change outfits)
- 
- Dbug on client 10 when a player from your vip list, train with exersice weapon (it is related to the fact that the player's name turns blue when training with exercise weapon)
- Dbug With Hilering (I imagine this happens because of the icons of npc hiring)
- A debug should also happen with the forge system in monsters, the influence, the new icons

dudantas committed:
Merge branch 'main' into multiprotocol

dudantas committed:
fix compilation error from conflict fix

Dengow-Matt commented:
Good night, I'm having conflicts with this pr

hriqueXimenes commented:
Hey @dudantas , I think the last commit in master for tibia 13.16 break this code

dudantas committed:
Merge branch 'main' into pr/516

dudantas commented:
Ok, fixed!

dudantas committed:
fix: some conflict errors

dudantas committed:
fix bugs

Fixed the relateds bugs:
Ooutfit cannot be changed with client 10 give debug, therefore in otcv8 it does not open the outfits panel (it only works with client 12, change outfits)
Debug on client 10 when a player from your vip list, train with exersice weapon (it is related to the fact that the player's name turns blue when training with exercise weapon)
Debug with Hilering (I imagine this happens because of the icons of npc hiring)
Debug should also happen with the forge system in monsters, the influence, the new icons

dudantas commented:
> Hello I have been testing, there are some old errors and I found a new one:
> 
> * The most critical error so far is that the outfit cannot be changed with client 10 give debug, therefore in otcv8 it does not open the outfits panel (it only works with client 12, change outfits)
> * Dbug on client 10 when a player from your vip list, train with exersice weapon (it is related to the fact that the player's name turns blue when training with exercise weapon)
> * Dbug With Hilering (I imagine this happens because of the icons of npc hiring)
> * A debug should also happen with the forge system in monsters, the influence, the new icons

![image](https://user-images.githubusercontent.com/8551443/234773394-91320ad5-3f19-4572-bed2-31f890e97e66.png)
![image](https://user-images.githubusercontent.com/8551443/234773412-7ff9b5a2-ba6a-4c98-b25d-f7cdd0cb408f.png)
![image](https://user-images.githubusercontent.com/8551443/234773439-326491df-f9c5-486b-a5ed-eb488a664abb.png)
![image](https://user-images.githubusercontent.com/8551443/234773529-0583b7bc-3a4b-4031-b1a8-4a863e70ff2f.png)

![image](https://user-images.githubusercontent.com/8551443/234778717-00e275e6-fa47-47e4-82ba-012939f1f16e.png)
![image](https://user-images.githubusercontent.com/8551443/234778739-a06241d2-56ad-4b8e-a5f8-8a5207139d3b.png)

Fixed all in https://github.com/opentibiabr/canary/pull/516/commits/76a3f85ccdb2526468671410890868730c71e1a1

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: send message if protocolOld is disabled and try login with client 1100

dudantas committed:
fix: debug with secondary spell

dudantas commented:
> Exevo gran mas frigo

Fixado: https://github.com/opentibiabr/canary/pull/516/commits/007eecd296b02d64fe89b8bb53a3906273d424e7
![image](https://user-images.githubusercontent.com/8551443/234787857-6ce804f0-9609-4579-8620-ad5c4e96980f.png)

dudantas committed:
fix: lastDay table on buy premium from store

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Merge branch 'main' into multiprotocol

dudantas committed:
fix lastday

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix typo

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
fix: lastday

dudantas committed:
conflict

dudantas committed:
Merge branch 'main' into pr/516

dudantas committed:
Merge branch 'multiprotocol' of https://github.com/marcosvf132/canary into pr/516

dudantas commented:
> Parece algo bom, contudo não vejo sentido em gastar tempo com isto, visto que muita coisa que há no 12 ainda nem finalizada foi. Fora que adequar o velho cliente a funções do 12 será um belo tempo gasto.

Aí que tá o lado bom, não precisa adaptar nada para o client 12. É basicamente ignorar os conteúdos novos se o jogador estiver logado no client antigo, usando o
```cpp
if (oldProtocol) {
    return
}
```

O pull request não vai atrapalhar em nada a gente, depois que estiver implementado, só vai ajudar a quem quer usar o client old.

Dengow-Matt commented:
> > Parece algo bom, contudo não vejo sentido em gastar tempo com isto, visto que muita coisa que há no 12 ainda nem finalizada foi. Fora que adequar o velho cliente a funções do 12 será um belo tempo gasto.
> 
> Aí que tá o lado bom, não precisa adaptar nada para o client 12. É basicamente ignorar os conteúdos novos se o jogador estiver logado no client antigo, usando o
> 
> ```c++
> if (oldProtocol) {
>     return
> }
> ```
> 
> O pull request não vai atrapalhar em nada a gente, depois que estiver implementado, só vai ajudar a quem quer usar o client old.

very good, it will help a lot
